indigo_api/models/timeline.py

```diff
@@ -0,0 +1,174 @@
+            setattr(self, k, v)
+
+
+class TimelineCommencementEvent(TimelineEvent):
+    subtype = None
+    by_work = None
+    in_full = None
+    date = None
+
+    def __init__(self, **kwargs):
+        self.type = 'commencement'
+        super().__init__(**kwargs)
+
+
+def describe_single_commencement(commencement, with_date=True, friendly_date=True):
+    description = TimelineCommencementEvent(subtype='single', related=commencement)
+
+    if commencement.note:
+        description.note = _('Note: %(note)s') % {'note': commencement.note}
+
+    if commencement.commencing_work:
+        description.by_frbr_uri = commencement.commencing_work.frbr_uri
+        description.by_title = commencement.commencing_work.title
+        description.by_work = commencement.commencing_work
+
+    commencement_date = commencement.date
+    if commencement_date:
+        description.date = commencement_date
+        if friendly_date:
+            commencement_date = date_format(commencement_date, 'j E Y')
+
+    if with_date and commencement_date:
+        # with a date and a commencing work
+        if commencement.commencing_work:
+            description.description = _('Commenced on %(date)s by') % {'date': commencement_date}
+
+        # with a date, without a commencing work
+        else:
+            description.description = _('Commenced on %(date)s') % {'date': commencement_date}
+
+    # without a date, with a commencing work
+    elif commencement.commencing_work:
+        description.description = _('Commenced by')
+
+    # without a date or a commencing work
+    else:
+        description.description = _('Commenced')
+
+    return description
+
+
+def get_timeline(work):
+    """ Returns a list of TimelineEvent objects, each describing a date on the timeline of a work.
+    """
+    entries = []
+
+    all_amendments = work.amendments.all()
+    all_commencements = work.commencements.all()
+    all_consolidations = work.arbitrary_expression_dates.all()
+
+    amendment_dates = [c.date for c in all_amendments]
+    commencement_dates = [c.date for c in all_commencements]
+    consolidation_dates = [c.date for c in all_consolidations]
+    other_dates = [work.assent_date, work.publication_date, work.repealed_date]
+    # don't include None
+    all_dates = [e for e in amendment_dates + commencement_dates + consolidation_dates + other_dates if e]
+    all_dates = set(all_dates)
+
+    # the initial date is the publication date, or the earliest of the consolidation and commencement dates, or None
+    initial = work.publication_date
+    if not initial and any(commencement_dates + consolidation_dates):
+        initial = min(d for d in commencement_dates + consolidation_dates if d)
+
+    for date in all_dates:
+        entry = TimelineEntry(date=date, initial=date == initial)
+        amendments = [a for a in all_amendments if a.date == date]
+        if len(amendments) > 1:
+            from indigo_api.models import Amendment
```

Comment:
if you're doing local imports, prefer to do them all at the start of the function